File: crates/lore-vm/src/ops/notification/subscribe.rs
//! `notification subscribe` operation — binds `lore::notification::subscribe`.
//!
//! Subscribes to repository push-event notifications. The subscription remains
//! active until a corresponding [`unsubscribe`](super::unsubscribe) call.

use crate::api::LoreApi;
use crate::collect::collect_events;
use crate::error::{LoreError, Result};

use lore::notification::LoreNotificationSubscribeArgs;
use serde::{Deserialize, Serialize};

/// Result returned on successful subscribe.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct SubscribeResult {}

/// Subscribe to repository notifications.
///
/// Calls the upstream `lore::notification::subscribe` in-process.
/// Returns `Ok(SubscribeResult)` when the subscribe succeeds.
pub async fn subscribe(api: &LoreApi) -> Result<SubscribeResult> {
    let (callback, rx) = collect_events();

    let status = lore::notification::subscribe(
        api.globals().build(),
        LoreNotificationSubscribeArgs {},
        callback,
    )
    .await;

    let stream = rx
        .await
        .map_err(|e| LoreError::CommandFailed(format!("event stream cancelled: {e}")))?;

    if !stream.is_ok() {
        return Err(LoreError::CommandFailed(stream.error.unwrap_or_else(
            || format!("subscribe failed with status {status}"),
        )));
    }

    Ok(SubscribeResult::default())
}

File: frontend/src/palette/manifest/notification/subscribe.ts
import type { OpManifest } from "../../types";

/**
 * Manifest entry for notification subscribe operation.
 *
 * Subscribes to repository push-event notifications. The subscription remains
 * active until a corresponding unsubscribe call.
 *
 * Phase 1 reference: no-arg op with JSON result.
 */
const manifest: OpManifest = {
  id: "notification.subscribe",
  domain: "notification",
  op: "subscribe",
  label: "Notification: Subscribe",
  description: "Subscribe to repository push-event notifications.",
  command: "notification_subscribe",
  args: [],
  resultKind: "json",
  keywords: ["notification", "subscribe", "events", "watch"],
};

export default manifest;
